### PR TITLE
Provide mechanism for skipping retries w/o acknowledging job

### PR DESCRIPTION
**Pull request by nyergler**
Prior to this commit a job could fail in one of two ways: it could
fail with an Unrecoverable error, in which case it'd be ack'd and
removed from the queue; or it could fail with another error and be
retried with exponential backoff.

There are some situations where neither is desirable; for example, you
might want to delay processing a job by manually re-Enqueuing it
without exponential backoff.

This commit introduces a new error, ErrDoNotRetry, and makes
ErrUnrecoverable wrap it. Rather than testing for equality to
ErrUnrecoverable when determining if we should retry, we check if the
error Is (or wraps) ErrDoNotRetry.

This allows the handler (or more likely handler middleware) to more
precisely control the processing of a job.

### CODE REVIEW

**Comment by taylorchu** [comment]
I am going to add a backoff function options, which can cover constant backoff.
But you can also add your handler middleware to enqueue, and simply return nil.

**Comment by nyergler** [comment]
That was what I tried first, @taylorchu; however, I found that unless I completely re-enqueued (with a new job ID, etc) that returning `nil` caused the job the be `ack`'d, which I wanted to avoid.

Do you suggest assigning new job IDs when re-enqueuing?

File: worker.go [hunk 1]
```diff
@@ -104,13 +104,18 @@ func (opt *JobOptions) Validate() error {
 }
 
 var (
+	// ErrDoNotRetry is returned if the job should not be retried;
+	// this may be because the job is unrecoverable, or because
+	// the handler has already rescheduled it.
+	ErrDoNotRetry = errors.New("do not retry")
```
Comment (taylorchu) [inline worker.go:110]: need to add a prefix "work: " to be clear for package user that it is from this package

File: worker.go [hunk 2]
```diff
@@ -104,13 +104,18 @@ func (opt *JobOptions) Validate() error {
...
+
 	// ErrQueueNotFound is returned if the queue is not yet
 	// defined with Register().
 	ErrQueueNotFound = errors.New("work: queue is not found")
 
 	// ErrUnrecoverable is returned if the error is unrecoverable.
 	// The job will be discarded.
-	ErrUnrecoverable = errors.New("work: permanent error")
+	ErrUnrecoverable = fmt.Errorf("work: permanent error%w", ErrDoNotRetry)
```
Comment (taylorchu) [inline worker.go:118]: need to be `: %w`. this printf pattern seems wrong.

File: worker.go [hunk 3]
```diff
@@ -378,6 +383,9 @@ func retry(queue Queue) HandleMiddleware {
 				})
 				return err
 			}
+			if err != ErrUnrecoverable {
```
Comment (taylorchu) [inline worker.go:386]: need to move this to be clearer
in err != nil block,
```
if err == ErrUnrecoverable {
  return nil // ack
}
if errors.Is(err, ErrDoNotRetry) {
   // don't ack and don't reenqueue
   return err
}
// ... reenqueue logic
return err
```

File: worker_test.go [hunk 4]
```diff
@@ -356,6 +356,15 @@ func TestRetry(t *testing.T) {
 	require.NoError(t, err)
 	require.Len(t, z, 0)
 
+	h = retrier(func(*Job, *DequeueOptions) error {
+		return fmt.Errorf("recoverable, but not retried%w", ErrDoNotRetry)
```
Comment (taylorchu) [inline worker_test.go:360]: need to fix printf format

**Review by taylorchu — CHANGES_REQUESTED** [review]
lgtm, added some suggestions.

**Comment by nyergler** [comment]
@taylorchu Apologies for dropping the ball on this, and thank you for shepherding it across the line. Much appreciated. 🙇‍♂️

**Pull request closed without merging**